pkg/apis/serving/v1alpha1/inference_graph.go

```diff
@@ -0,0 +1,252 @@
+
+// InferenceRouter defines the router for each InferenceGraph node with one or multiple models
+// and where it routes to as next step
+//
+// ```yaml
+// kind: InferenceGraph
+// metadata:
+//   name: canary-route
+// spec:
+//   nodes:
+//     mymodel:
+//       routerType: Splitter
+//       routes:
+//       - service: mymodel1
+//         weight: 20
+//       - service: mymodel2
+//         weight: 80
+// ```
+//
+// ```yaml
+// kind: InferenceGraph
+// metadata:
+//   name: abtest
+// spec:
+//   nodes:
+//     mymodel:
+//       routerType: Switch
+//       routes:
+//       - service: mymodel1
+//         condition: "{ $input.userId == 1 }"
+//       - service: mymodel2
+//         condition: "{ $input.userId == 2 }"
+// ```
+//
+// Scoring a case using a model ensemble consists of scoring it using each model separately,
+// then combining the results into a single scoring result using one of the pre-defined combination methods.
+//
+// Tree Ensemble constitutes a case where simple algorithms for combining results of either classification or regression trees are well known.
+// Multiple classification trees, for example, are commonly combined using a "majority-vote" method.
+// Multiple regression trees are often combined using various averaging techniques.
+// e.g tagging models with segment identifiers and weights to be used for their combination in these ways.
+// ```yaml
+// kind: InferenceGraph
+// metadata:
+//   name: ensemble
+// spec:
+//   nodes:
+//     transformer:
+//       routes:
+//       - service: feast
+//       nextRoutes:
+//       - nodeName: ensembleModel
+//         data: "{ $output }"
+//     ensembleModel:
+//       routerType: Ensemble
+//       routes:
+//       - service: sklearn-model
+//       - service: xgboost-model
+// ```
+//
+// Scoring a case using a sequence, or chain, of models allows the output of one model to be passed in as input to subsequent models.
+// ```yaml
+// kind: InferenceGraph
+// metadata:
+//   name: model-chainer
+// spec:
+//   nodes:
+//     mymodel-s1:
+//       routes:
+//       - service: mymodel-s1
+//       nextRoutes:
+//       - nodeName: mymodel-s2
+//         data: "{ $output }"
+//     mymodel-s2:
+//       routes:
+//       - service: mymodel-s2
+// ```
+//
+// In the flow described below, the pre_processing node base64 encodes the image and passes it to two model nodes in the flow.
+// The encoded data is available to both these nodes for classification. The second node i.e. dog-breed-classification takes the
+// original input from the pre_processing node along-with the response from the cat-dog-classification node to do further classification
+// of the dog breed if required.
+// ```yaml
+// kind: InferenceGraph
+// metadata:
+//   name: dog-breed-classification
+// spec:
+//   nodes:
+//     top-level-classifier:
+//       routes:
+//       - service: cat-dog-classifier
+//       nextRoutes:
+//         nodeName: breed-classifier
+//         data: "{ $input }"
```

Comment:
Syntax for this looks inconsistent with other code/docs. (`{ $input }` vs `$request`)